Block/SystemInfo.php

```diff
@@ -0,0 +1,105 @@
+use Magento\Backend\Block\Template\Context;
+use Magento\Config\Block\System\Config\Form\Field;
+use Magento\Framework\Data\Form\Element\AbstractElement;
+
+class SystemInfo extends Field
+{
+    /**
+     * @var ApplicationService
+     */
+    private $applicationService;
+
+    /**
+     * @var Data
+     */
+    private $helper;
+
+    public function __construct(ApplicationService $applicationService, Data $helper, Context $context, array $data = [])
+    {
+        parent::__construct($context, $data);
+
+        $this->applicationService = $applicationService;
+        $this->helper = $helper;
+    }
+
+    public function render(AbstractElement $element): string
+    {
+        $infoMessages = [];
+        $successMessages = [];
+        $warningMessages = [];
+        $errorMessages = [];
+
+        $systemInfo = sprintf(
+            'Magento Comfino %s, Magento %s, PHP %s, web server %s, database %s',
+            $this->helper->getModuleVersion(),
+            $this->helper->getShopVersion(),
+            PHP_VERSION,
+            $_SERVER['SERVER_SOFTWARE'],
+            $this->helper->getDatabaseVersion()
+        );
+
+        $bodyHtml = '<p><label class="label">' . __('System information') . '</label></p>';
+
+        $infoMessages[] = $systemInfo;
+
+        if ($this->helper->isSandboxEnabled()) {
+            $warningMessages[] = __('Developer mode is active. You are using test environment.');
+
+            if (!empty($this->helper->getApiKey())) {
+                if ($this->applicationService->isShopAccountActive()) {
+                    $successMessages[] = __('Test account is active.');
+                } else if (count($this->applicationService->getLastErrors())) {
+                    $errorMessages = array_merge($errorMessages, $this->applicationService->getLastErrors());
+
+                    if ($this->applicationService->getLastResponseCode() === 401) {
+                        $errorMessages[] = __('Invalid test API key.');
+                    }
+                } else {
+                    $warningMessages[] = __('Test account is not active.');
+                }
+            } else {
+                $errorMessages[] = __('Test API key not present.');
+            }
+        } elseif (!empty($this->helper->getApiKey())) {
+            $successMessages[] = __('Production mode is active.');
+
+            if ($this->applicationService->isShopAccountActive()) {
+                $successMessages[] = __('Production account is active.');
+            } else if (count($this->applicationService->getLastErrors())) {
+                $errorMessages = array_merge($errorMessages, $this->applicationService->getLastErrors());
+
+                if ($this->applicationService->getLastResponseCode() === 401) {
+                    $errorMessages[] = __('Invalid production API key.');
+                }
+            } else {
+                $warningMessages[] = __('Production account is not active.');
+            }
```

Comment:
Zawsze mam problem z czytaniem takiej zgnieżdzonej ifologi

Reply:
To jest posranie - nie ukrywam. :-D Mi też się nie podoba. Może kiedyś coś bardziej zwartego wymyślę, ale to już raczej w ramach wydzielenia wspólnej logiki backendu wtyczek do zewnętrznej biblioteki pobieranej composerem, tak jak teraz to mamy już zrobione z frontem. To też prosi się o wydzielenie na zewnątrz, bo to kolejna duplikacja logiki w dużych ilościach.